mkosi

```diff
@@ -415,8 +415,22 @@ def reuse_cache_image(args, workspace, run_build_script, for_cache):
         with source:
             f = tempfile.NamedTemporaryFile(dir = os.path.dirname(args.output), prefix='.mkosi-')
             output.append(f)
+
+            # So on one hand we want CoW off, since this stuff will
+            # have a lot of random write accesses. On the other we
+            # want the copy to be snappy, hence we do want CoW. Let's
+            # ask for both, and let the kernel figure things out:
+            # let's turn off CoW on the file, but start with a CoW
+            # copy. On btrfs that works: the initial copy is made as
+            # CoW but later changes do not result in CoW anymore.
+
             disable_cow(f.name)
-            shutil.copyfileobj(source, f)
+            try:
+                _reflink(source.fileno(), f.fileno())
+            except OSError as e:
+                if e.errno not in {errno.EXDEV, errno.EOPNOTSUPP}:
+                    raise
+                shutil.copyfileobj(source, f)
```

Comment:
Couldn’t this use the `_copy_file` function? It looks like it does mostly the same thing, and I’m not sure if the slight differences are intended.

Reply:
copy_file() takes a path, _copy_file() takes fd integers, but we need python file objects here...

But you are right, we should probably split this into a function of its own. Will do.